doc/port-profile-init/port-profile-init-design.md

```diff
@@ -0,0 +1,179 @@
+
+<p align=left>
+<img src="img/old_flow.svg" alt="Figure 1. Old Flow">
+</p>
+
+
+<p align=left>
+<img src="img/new_flow.svg" alt="Figure 2. New Flow">
+</p>
+
+
+The above sequence diagram shows that a lot of calls to SAI and SDK APIs are being saved.
+In terms of Fast Reboot, when system is going up, it can save a lot of time and not challenge the 30 seconds restriction.
+
+Notes:
+
+* Multi-Asic is covered for these changes. In Multi-Asic environment, there are multiple instances of swss and Redis DB and each DB has it's own PORT table. Hence, port configuration is per Asic.
+
+* When create_ports() API is called with an empty list of ports, SONiC expects SAI to return any status except of SAI_STATUS_NOT_IMPLEMENTED or SAI_STATUS_NOT_SUPPORTED. This call will be performed in order to validate the API is supported or not.
+
+#### Initialization Flow
+
+As part of PortsOrch initialization we query for all pre-configured ports in SAI and put that in a list for the later comparison logic.
+
+Now that SAI will no longer initialize all ports from SAI profile file, this query is not relevant.
+It will no longer be part of PortsOrch initialization and will only be done in case bulk option is not supported.
+
+2 checks will be performed in initialization flow.
+1. Is bulk option supported? It will be checked using the following flow:
+
+    * If sai_port_api->create_ports() is NULL --> Not supported.
+    * If not NULL, call it (with an empty list of ports).
+    * If return status is SAI_STATUS_NOT_IMPLEMENTED or SAI_STATUS_NOT_SUPPORTED --> Not supported.
+    * Otherwise, supported.
+
+2. Are ports configured in SAI?
+
+    In order to check this, we will query the number of ports created in SAI using sai_switch_api->get_switch_attribute() API, while attr id is SAI_SWITCH_ATTR_PORT_NUMBER.
+
+    If ports number is greater than 0, fallback to legacy flow.
+
+
+#### doPortTask Flow
+
+Currently, in case a new port was added to PORT table in APP DB, PortsOrch compares the new port configurations in APP DB, to the pre configured port attributes queried from SAI. In case of a mismatch, it will remove the configured port and recreate port with attributes taken from APP DB.
+
+According to new logic, in case BULK option is supported, the function will wait for PortConfigDone flag to be raised. Meaning, all ports from Config DB are now in APP DB.
+
+Then, per each port, it will add it's attributes to sai_attribute_t list. After all attributes are added, sai_port_api->create_ports() API will be called.
+
+\* Creating Host Interfaces and setting SAI attributes will be done one by one as beofre.
+
+
+### SAI API
+
+https://github.com/opencomputeproject/SAI/pull/1460/files SAI PR introduces new SAI APIs to support bulk creation.
+
+sai_port module will now have the following APIs:
+
+```
+sai_bulk_object_create_fn              create_ports;
+sai_bulk_object_remove_fn              remove_ports;
+sai_bulk_object_set_attribute_fn       set_ports_attribute;
```

Comment:
Please see if we can use bulk set_ports_attribute instead of individual set attribute API.

Reply:
Done